native/src/main/scala/sqs4s/auth/CredProvider.scala

```diff
@@ -0,0 +1,159 @@
+package sqs4s.auth
+
+import java.util.concurrent.TimeUnit
+
+import cats.Applicative
+import cats.implicits._
+import cats.effect.implicits._
+import cats.effect.concurrent.Ref
+import cats.effect.{Clock, Concurrent, Sync, Timer}
+import fs2._
+import org.http4s.Method.{GET, PUT}
+import org.http4s.circe.CirceEntityDecoder._
+import org.http4s.client.Client
+import org.http4s.client.dsl.Http4sClientDsl
+import org.http4s.implicits._
+import org.http4s.{Header, Response, Status}
+import sqs4s.auth.errors._
+
+import scala.concurrent.duration._
+
+trait CredProvider[F[_]] {
+  def accessKey: F[String]
+  def secretKey: F[String]
+}
+
+trait SessionCredProfiler[F[_]] {
+  def sessionToken: F[String]
+}
+
+trait TemporaryCredProvider[F[_]]
+    extends CredProvider[F]
+    with SessionCredProfiler[F] {
+  def refresh: F[Credential]
+}
+
+case class BasicCredProvider[F[_]: Applicative](
+  accessK: String,
+  secretK: String
+) extends CredProvider[F] {
+  def accessKey: F[String] = accessK.pure[F]
+  def secretKey: F[String] = secretK.pure[F]
+}
+
+class InstanceProfileCredProvider[F[_]: Sync: Clock: Timer](
```

Comment:
I would call this `RoleCredProvider` since this should work also in **ECS**, **EKS**, and even **Lambda**.

Also, in the future, we could add an `EnvCredProvider` which looks for the access tokens in the environment and another one `ConfigFileCredProvider` that searches for the access tokens in  the `.aws/credentials` file.

Reply:
the naming is to mirror what AWS SDK provide: https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/InstanceProfileCredentialsProvider.html. My plan is to provider Env and Java properties next, and have a top level one to look for all available options just like the SDK.

Reply:
> the naming is to mirror what AWS SDK provide:

Ok makes sense. I guess they named like that from a beginning and couldn't change it later, but that is just speculation.

> have a top level one to look for all available options just like the SDK.


Amazing!